src/lex.cpp

```diff
@@ -709,9 +711,33 @@ void lexer::parse_number() {
   this->input_ = input;
 }
 
+bool has_trailing_underscore = false;
+const char8* lexer::parse_underscore(const char8* input) noexcept {
+  while (is_digit(*input)) {
+    has_trailing_underscore = false;
+    input += 1;
+    if (*input == '_') {
+      input += 1;
+      has_trailing_underscore = true;
+      if (*input == '_') {
+        has_trailing_underscore = false;
+        // TODO error_reporter: SyntaxError: Only one underscore is allowed in numeric separator
+      }
+    }
+  }
+  if (has_trailing_underscore == true) {
+    // TODO error_reporter: SyntaxError: Numeric separators are not allowed at the end of numeric literals
+  }
+  return input;
+}
+
 const char8* lexer::parse_decimal_digits(const char8* input) noexcept {
```

Comment:
Discussed in person: parse_decimal_digits and parse_underscores are doing the same thing. Merge them into one function.